### ISSUE TITLE
[fbpick-coarse] Replace tiled coarse inference with global-anchor inference  Issue body

### ISSUE THREAD

**Issue by ozw4** [ozw4/seisai#30]
Background

fbpick-coarse の train / validation dataset は global-anchor resize mode に移行済み。

現在の coarse training / validation は以下の固定契約になっている。

input mode: global_anchor_resize
input shape:  (3, 256, 2048)
target shape: (1, 256, 2048)

channels:
  0: waveform
  1: offset_ch
  2: time_ch

一方で、raw coarse inference はまだ legacy tiled path を使っている。

現状の inference path は以下の構成。

InferenceGatherWindowsDataset
collate_pad_w_right
TiledWConfig
iter_infer_loader_tiled_w
tile_w / overlap_w / tiles_per_batch
H-window / W-tile stitching

このままだと、global-anchor coarse checkpoint を train / validation しても、raw inference 側では tiled window 前提の入力に流れる可能性がある。

この issue では、coarse raw inference を tiled 方式から global-anchor 方式へ置き換える。

Goal

run_fbpick_coarse_infer.py / coarse/infer.py の inference path を、gather ごとに固定サイズ 3×256×2048 の global-anchor input を作って forward する方式へ変更する。

新しい inference flow:

SEG-Y full gather
→ gap-aware center anchor selection
→ selected anchor traces: Hc=256
→ time resampling: Wc=2048
→ input: (3, 256, 2048)
→ model forward once per gather
→ anchor-level coarse prediction
→ 元 trace / 元 sample 座標へ復元
→ .coarse.npz 保存

この issue では、raw inference の global-anchor 化と .coarse.npz output contract 維持まで扱う。

Prerequisites

以下は完了済み前提。

[fbpick-coarse] Add global-anchor 256×2048 config and checkpoint contract
[dataset] Add full-gather sampling support
[fbpick-coarse] Implement gap-aware trace anchor selection
[fbpick-coarse] Implement time resampling and FB coordinate projection
[fbpick-coarse] Replace train/validation datasets with global-anchor datasets

利用可能な想定 API:

TraceSubsetSampler.draw_full_gather
select_trace_anchors / select_gap_aware_trace_anchors
resample_waveform_time_axis
build_coarse_time_grid
project_coarse_indices_to_raw_time

実際の関数名は既存実装に合わせる。

Scope
1. Add global-anchor raw inference dataset

対象候補:

packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/build_dataset.py

既存の build_raw_infer_dataset は InferenceGatherWindowsDataset を返している。
これを global-anchor raw inference dataset に置き換える。

想定 class:

GlobalAnchorCoarseRawInferDataset

または既存命名規則に合わせる。

dataset の責務:

1. raw SEG-Y file info を読む
2. primary gather 全体を draw_full_gather で取得する
3. full gather の trace indices / offsets / headers を取得する
4. gap-aware center anchor selection を行う
5. anchor trace waveform を読み込む
6. waveform を Wc=2048 に time resample する
7. waveform を per-trace standardize する
8. offset_ch / time_ch を BuildPlan 経由で作る
9. input tensor と restoration metadata を返す

raw inference なので target は返さない。

期待 shape:

sample["input"].shape == (3, 256, 2048)
2. Use deterministic center anchors for inference

raw inference では anchor selection を必ず deterministic にする。

anchor_mode = "center"

期待:

- 同じ SEG-Y / config / checkpoint から同じ anchor mapping が返る
- rng に依存しない
- gap をまたいだ anchor bin を作らない
- pad row は trace_valid=False
3. Preserve metadata needed for output restoration

dataset item には、anchor prediction を元 trace / 元 sample 座標に戻すための metadata を含める。

最低限必要な metadata:

source_file
raw_trace_indices
anchor_raw_indices
anchor_source_pos
anchor_offsets_m
trace_valid
segment_id
anchor_bin_start_pos
anchor_bin_stop_pos
raw_time_len
coarse_time_len
dt_sec
dt_eff_sec
raw_to_coarse_factor
coarse_to_raw_factor
offsets_m for full gather
ffid_values
chno_values
n_traces
n_samples_orig

DataLoader collate が難しい場合、初期実装では batch_size=1 を必須にしてよい。

その場合は config validation で明確に落とす。

global-anchor coarse inference currently requires infer.batch_size == 1
4. Remove tiled inference path from coarse inference

対象候補:

packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/infer.py

global-anchor path から以下を外す。

InferenceGatherWindowsDataset
collate_pad_w_right
TiledWConfig
iter_infer_loader_tiled_w
tile_w
overlap_w
tiles_per_batch
H-window stitching
W-tile stitching

新しい inference loop は通常の model forward にする。

概念:

for batch in loader:
    x = batch["input"].to(device)  # (B, 3, 256, 2048)

    with torch.no_grad():
        logits = model(x)          # (B, 1, 256, 2048)

shape validation:

logits.shape == (B, 1, 256, 2048)

不一致なら即エラー。

5. Extract anchor-level predictions

model output から anchor row ごとの prediction を得る。

logits[b, 0].shape == (256, 2048)

算出:

anchor_pick_j:
  argmax over time axis
  shape: (256,)

anchor_pmax:
  softmax over time axis の max probability
  shape: (256,)

pad row は trace_valid=False で無視する。

6. Project coarse time predictions back to raw sample coordinates

anchor_pick_j は coarse grid 上の sample index なので、元 SEG-Y の sample index に戻す。

必ず既存 utility を使う。

project_coarse_indices_to_raw_time

変換結果:

anchor_pick_i_raw.shape == (256,)

valid row のみ使う。

7. Restore anchor predictions to full raw trace count

anchor row で得た prediction を、元 gather の全 trace 数へ戻す。

方針:

- segment 内だけで補間する
- offset gap をまたいで補間しない
- segment 内 anchor が 1 本だけの場合は、その segment 全体に同じ pick を割り当てる
- coarse_pmax も segment 内で補間または nearest copy する

推奨:

trace order position に対する np.interp

例:

pred_i_segment = np.interp(
    raw_positions_in_segment,
    anchor_positions_in_segment,
    anchor_pick_i_raw_segment,
)

最終 output:

coarse_pick_i.shape == (n_traces,)
coarse_pmax.shape == (n_traces,)

最後に必ず:

coarse_pick_i = np.rint(coarse_pick_i).clip(0, n_samples_orig - 1).astype(np.int32)
coarse_pick_t_sec = coarse_pick_i.astype(np.float32) * np.float32(dt_sec)

coarse_pick_t_sec は coarse_pick_i から再計算し、整合性を保つ。

8. Preserve .coarse.npz output contract

既存 downstream の physics / fine stage が壊れないように、.coarse.npz の外部 contract は維持する。

保存する主要 field:

dt_sec
n_samples_orig
n_traces
ffid_values
chno_values
offsets_m
trace_indices
coarse_pick_i
coarse_pick_t_sec
coarse_pmax
coarse_prob_summary
lineage

内部 inference が global-anchor になっても、保存値は元 SEG-Y 座標に戻したものにする。

9. Enforce checkpoint metadata validation

global-anchor inference では checkpoint metadata が以下に一致することを確認する。

coarse_input_mode == "global_anchor_resize"
coarse_trace_len == 256
coarse_time_len == 2048
coarse_in_chans == 3

metadata 欠落 checkpoint や legacy tiled coarse checkpoint は、明確なエラーで停止する。

10. Update infer config example

対象候補:

examples/config_infer_fbpick_coarse.yaml

global-anchor inference では tiled-specific config を使わない。

削除または deprecated 扱いにする項目:

overlap_h
tile_w
overlap_w
tiles_per_batch

新しい例:

coarse:
  input_mode: global_anchor_resize

transform:
  trace_len: 256
  time_len: 2048
  standardize_eps: 1.0e-8

trace_anchor:
  gap_ratio: 5.0
  min_gap_m: null
  infer_mode: center

infer:
  ckpt_path: REPLACE_ME.pt
  device: auto
  batch_size: 1
  num_workers: 0
  amp: false
  use_tqdm: true
Out of scope

この issue では以下は行わない。

- train / validation dataset の変更
- gap-aware anchor selection utility の再設計
- time resampling utility の再設計
- model architecture の変更
- physics / fine stage の変更
- viewer の大規模変更
- unrelated formatting cleanup
Acceptance criteria
raw coarse inference dataset が global-anchor input を返す。
raw inference input shape が (3, 256, 2048) である。
DataLoader batch input shape が (B, 3, 256, 2048) である。
raw inference dataset は draw_full_gather を使い、H-window crop を行わない。
raw inference dataset は mode="center" の gap-aware anchor selection を使う。
raw inference dataset は deterministic である。
coarse inference loop から tiled-W inference 依存が外れている。
iter_infer_loader_tiled_w, TiledWConfig, tile_w, overlap_w, tiles_per_batch が global-anchor path で使われない。
model forward は gather ごとに global input へ一度実行される。
logits shape (B, 1, 256, 2048) が検証される。
anchor-level anchor_pick_j / anchor_pmax が取得できる。
anchor_pick_j が raw sample index に復元される。
anchor prediction が segment 内だけで full trace count へ補間される。
gap をまたいだ補間が行われない。
.coarse.npz は既存 downstream contract を維持する。
coarse_pick_i, coarse_pick_t_sec, coarse_pmax の shape は (n_traces,)。
coarse_pick_t_sec == coarse_pick_i * dt_sec の整合性が保たれる。
checkpoint metadata validation が infer path で実行される。
legacy tiled coarse checkpoint は global-anchor inference で明確に失敗する。
example infer config が global-anchor inference 前提に更新されている。
unit tests / smoke tests が追加または更新されている。
既存 tests が通る。
Suggested tests
1. Raw infer dataset returns fixed shape

期待:

sample["input"].shape == (3, 256, 2048)
2. Raw infer dataset has no target

期待:

"target" not in sample

または既存 convention に合わせて target is None。

3. Raw infer dataset is deterministic

同じ item を複数回取得する。

期待:

anchor_source_pos が完全一致
input が完全一致または数値誤差範囲内で一致
4. Tiled inference dependencies are not used

global-anchor inference config で infer を実行し、以下が呼ばれないことを確認する。

iter_infer_loader_tiled_w
TiledWConfig

可能なら monkeypatch で、呼ばれたら fail するようにする。

5. Model forward shape

dummy model で forward input shape を確認する。

期待:

x.shape == (B, 3, 256, 2048)
6. Invalid logits shape fails

dummy model が不正 shape を返す。

(B, 1, 128, 2048)

期待:

ValueError
7. Anchor prediction extraction

time-axis が最大になる dummy logits を返す model を使う。

期待:

anchor_pick_j == 2047 for valid rows
anchor_pmax.shape == (256,)
8. Coarse-to-raw sample projection

既知の raw_time_len / coarse_time_len を使う。

例:

raw_time_len = 6016
coarse_time_len = 2048
anchor_pick_j = [0, 2047]

期待:

anchor_pick_i_raw = [0, 6015]
9. Segment-wise interpolation does not cross offset gaps

人工 metadata を使う。

segment 0: traces 0..9
segment 1: traces 10..19
gap between 9 and 10

期待:

segment 0 の prediction は segment 0 anchors のみから補間
segment 1 の prediction は segment 1 anchors のみから補間
gap をまたいだ np.interp が行われない
10. .coarse.npz contract smoke test

小さい SEG-Y fixture と dummy model で inference を実行する。

期待:

output file exists
coarse_pick_i.shape == (n_traces,)
coarse_pick_t_sec.shape == (n_traces,)
coarse_pmax.shape == (n_traces,)
coarse_pick_t_sec == coarse_pick_i * dt_sec
11. Checkpoint metadata validation

metadata 欠落 checkpoint または legacy tiled checkpoint を使う。

期待:

ValueError

message に以下を含む。

expected coarse_input_mode='global_anchor_resize'
Implementation notes

推奨する分割:

build_dataset.py
  - GlobalAnchorCoarseRawInferDataset
  - build_raw_infer_dataset を global-anchor 版に変更

infer.py
  - tiled imports を削除
  - DataLoader は default collate or simple custom collate
  - model forward
  - logits shape validation
  - anchor prediction extraction
  - raw coordinate restoration
  - segment-wise full trace interpolation
  - save_coarse_npz

tests
  - raw dataset shape
  - deterministic center anchors
  - no tiled dependency
  - coordinate restoration
  - .coarse.npz smoke

座標変換式は infer loop に直接書かず、既存 utility を使う。

project_coarse_indices_to_raw_time

trace 方向の復元も pure function として分けると test しやすい。

候補:

restore_anchor_predictions_to_full_traces(...)
